### PR TITLE
Akimbo alt fire

### PR DESCRIPTION
**Pull request by MiniMeatwad**
## About The Pull Request

Alternating fire with a mob boolean to determine which gun fires relative to the active hand when you have two guns.

## Why It's Good For The Game

Akimbo alternating fire

No more double fire killing balance.

## Changelog
:cl:
code: Akimbo alternates fire between hands now instead of firing both guns at once.
/:cl:

In-game:
https://user-images.githubusercontent.com/22848292/134723650-820bb3ba-fe37-4325-8c65-9da3f459f368.mp4
https://user-images.githubusercontent.com/22848292/134723652-b2186b9a-0cc6-45f3-b3a8-b37983e6d68d.mp4

### CODE REVIEW

**Comment by EvilMelonTheSecond** [comment]
Now this is proper balance. No number adjustment, no hard blocks AND an upside of increased magazine capacity in exchange.

**Comment by mathis650** [comment]
finally an intelligent solution to akimbo

**Comment by Dolth** [comment]
If that means you can shoot guns individually then it returns to what CM was long time ago therefor one handed bursts will have overlapped bursts so you still technically can shoot both at the same time on burst (just saying, it's not an issue or anything :P)

**Comment by MiniMeatwad** [comment]
> 
> 
> If that means you can shoot guns individually then it returns to what CM was long time ago therefor one handed bursts will have overlapped bursts so you still technically can shoot both at the same time on burst (just saying, it's not an issue or anything :P)

Sounds about right. If it's abusable and not fun to play with I could look into it later

File: code/modules/projectiles/gun_system.dm [hunk 1]
```diff
@@ -961,7 +961,10 @@ and you're good to go.
 /obj/item/weapon/gun/proc/able_to_fire(mob/user)
 	if(!user || user.stat != CONSCIOUS || user.lying_angle)
 		return
-
+	if(dual_wield && gun_user.get_active_held_item() == src & gun_user.shoot_inactive_hand)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:964]:
```suggestion
	if(dual_wield && gun_user.get_active_held_item() == src && gun_user.shoot_inactive_hand)
```

File: code/modules/projectiles/gun_system.dm [hunk 2]
```diff
@@ -961,7 +961,10 @@ and you're good to go.
...
+		return FALSE
+	if(dual_wield && gun_user.get_inactive_held_item() == src & !gun_user.shoot_inactive_hand)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:966]:
```suggestion
	if(dual_wield && gun_user.get_inactive_held_item() == src && !gun_user.shoot_inactive_hand)
```

File: code/modules/projectiles/gun_system.dm [hunk 3]
```diff
@@ -800,7 +799,9 @@ and you're good to go.
 	//^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 	shots_fired++
-
+	// Dumb akimbo code
+	if(dual_wield && (!gun_user.get_active_held_item().current_mag || gun_user.get_active_held_item().current_mag.current_rounds <= 0))
+		gun_user.shoot_inactive_hand = TRUE
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:804]: You want to do the same for when gun_user.shoot_inactive_hand is FALSE, to make a cycle

File: code/modules/projectiles/gun_system.dm [hunk 4]
```diff
@@ -817,7 +818,6 @@ and you're good to go.
 			sentry_battery.forceMove(get_turf(src))
 			sentry_battery.charge = 0
 			sentry_battery = null
-	return TRUE
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:820]: what is this?

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_system.dm (hunk 4 above)
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm:820]: mistake

File: code/modules/projectiles/gun_system.dm (hunk 3 above)
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm:804]: Yea but I am still working on it and can't make it work, cause i am bad

File: code/modules/projectiles/gun_system.dm [hunk 5]
```diff
@@ -800,9 +800,10 @@ and you're good to go.
 
 	shots_fired++
 	// Dumb akimbo code
-	if(dual_wield && gun_user.shoot_inactive_hand && (gun_user.get_active_held_item().current_mag || gun_user.get_active_held_item().current_mag.current_rounds > 0))
+	var/obj/item/weapon/gun/active_gun = gun_user.get_active_held_item() // Don't use this outside of dual_wield since there are no checks for if it's not a gun else where
+	if(dual_wield && gun_user.shoot_inactive_hand && active_gun?.current_mag.current_rounds > 0)
 		gun_user.shoot_inactive_hand = FALSE
-	if(dual_wield && !gun_user.shoot_inactive_hand && (gun_user.get_inactive_held_item().current_mag || gun_user.get_inactive_held_item().current_mag.current_rounds > 0))
+	if(dual_wield && !gun_user.shoot_inactive_hand && (!active_gun?.current_mag || active_gun.current_mag.current_rounds <= 0))
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:806]: Nop, you want to do exactly the same thing that you are doing line 803

cast the inactive_gun, and then && inactive_gun?.current_mag.current_rounds > 0

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_system.dm [hunk 6]
```diff
@@ -817,8 +817,16 @@ and you're good to go.
 			sentry_battery.forceMove(get_turf(src))
 			sentry_battery.charge = 0
 			sentry_battery = null
+	// Akimbo code
+	var/obj/item/weapon/gun/active_gun = gun_user.get_active_held_item() // Don't use active or inactive_gun outside of dual_wield since there are no checks for if it's not a gun else where
+	var/obj/item/weapon/gun/inactive_gun = gun_user.get_inactive_held_item()
+	if(!dual_wield || dual_wield && gun_user.shoot_inactive_hand && active_gun.current_mag?.current_rounds > 0)
+		gun_user.shoot_inactive_hand = FALSE
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:824]: setting back shoot_inactive hand should be done when dual wielding is set.

remove the !dual_wield here, reset shoot_inactive_hand at line 628

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_system.dm [hunk 7]
```diff
@@ -962,6 +970,10 @@ and you're good to go.
 	if(!user || user.stat != CONSCIOUS || user.lying_angle)
 		return
 
+	if(dual_wield && gun_user.get_active_held_item() == src && gun_user.shoot_inactive_hand)
+		return FALSE
+	if(dual_wield && gun_user.get_inactive_held_item() == src && !gun_user.shoot_inactive_hand)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:975]:
```suggestion
	else if(dual_wield && gun_user.get_inactive_held_item() == src && !gun_user.shoot_inactive_hand)
```

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_system.dm (hunk 6 above)
Comment (BraveMole) [reply code/modules/projectiles/gun_system.dm:824]: Also, very important, you must apply a delay on the next gun that will shoot. Else this is just normal akimbo but slightly slowler

**Review by BraveMole — CHANGES_REQUESTED** [review]
You'll also have to think what you'll do with other fire mode.

This is to kill akimbo pistol being broke, which is good, but what about T90 akimbo? It's shit, it could beneficiate from this somehow

File: code/modules/projectiles/gun_system.dm (hunk 6 above)
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm:826]: Shoot inactive hand reset on line 628

File: code/modules/projectiles/gun_system.dm [hunk 8]
```diff
@@ -687,6 +687,7 @@ User can be passed as null, (a gun reloading itself for instance), so we need to
 	set_target(null)
 	windup_checked = WEAPON_WINDUP_NOT_CHECKED
 	dual_wield = FALSE
+	gun_user.shoot_inactive_hand = FALSE
```
[the remark this answers is not in the document]
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm]: Okay

**Comment by MiniMeatwad** [comment]
How do I go about adding a delay to a gun?

**Comment by BraveMole** [comment]
> How do I go about adding a delay to a gun? I can only think of a way that involves adding new vars and procs cause I don't know gun code

look at the last_fired var and gun_on_cooldown proc

File: code/modules/projectiles/gun_system.dm [hunk 9]
```diff
@@ -421,6 +422,18 @@
...
+	if(!gun_user)
+		return
+	if(!isgun(gun_user.get_active_held_item()) || !isgun(gun_user.get_inactive_held_item()))
+		gun_user.shoot_inactive_hand = FALSE
+		return
+	if(gun_user.shoot_inactive_hand && gun_user.get_active_held_item().current_mag?.current_rounds && !gun_user.get_active_held_item().current_mag.current_rounds <= 0)
+		gun_user.shoot_inactive_hand = FALSE
+	if(!gun_user.shoot_inactive_hand && gun_user.get_inactive_held_item().current_mag?.current_rounds && !gun_user.get_inactive_held_item().current_mag.current_rounds <= 0)
+		gun_user.shoot_inactive_hand = TRUE
+
```
Comment (MiniMeatwad) [inline code/modules/projectiles/gun_system.dm:436]: I mean, it works...

**Comment by MiniMeatwad** [comment]
If you want to undraft and merge it now, even though the code isn't really of high quality, go for it.

File: code/modules/mob/mob_defines.dm [hunk 10]
```diff
@@ -27,6 +27,7 @@
 	var/track_blood = 0
 	var/feet_blood_color
 	var/datum/skills/skills
+	var/shoot_inactive_hand = FALSE
```
Comment (BraveMole) [inline code/modules/mob/mob_defines.dm:30]: need autodoc
///In akimbo firing, if the next hand shooting will be the inactive hand

File: code/modules/projectiles/gun_system.dm [hunk 11]
```diff
@@ -84,6 +84,8 @@
 
 
 	//Energy Weapons
+	var/obj/item/cell/cell						//Energy guns use cells instead of magazines.
+	var/cell_type = /obj/item/cell				//Default cell type, 1000 power.
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:88]: no need to move these vars

File: code/modules/projectiles/gun_system.dm [hunk 12]
```diff
@@ -228,6 +230,7 @@
 		QDEL_NULL(current_mag)
 	if(muzzle_flash)
 		QDEL_NULL(muzzle_flash)
+	QDEL_NULL(cell)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:233]: no need to move this

File: code/modules/projectiles/gun_system.dm [hunk 13]
```diff
@@ -652,6 +655,30 @@ User can be passed as null, (a gun reloading itself for instance), so we need to
 		SEND_SIGNAL(gun_user, COMSIG_MOB_ATTACHMENT_FIRED, target, src, master_gun)
 	gun_user?.client?.mouse_pointer_icon = 'icons/effects/supplypod_target.dmi'
 
+/obj/item/weapon/gun/proc/set_shoot_inactive_hand(mob/user, firing) // Handles akimbo
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:658]: autodoc.

File: code/modules/projectiles/gun_system.dm [hunk 14]
```diff
@@ -652,6 +655,30 @@ User can be passed as null, (a gun reloading itself for instance), so we need to
...
+		return
+	
+	var/obj/item/weapon/gun/active_gun = user.get_active_held_item()
+	var/obj/item/weapon/gun/inactive_gun = user.get_inactive_held_item()
+	if(firing) // Apply akimbo delay
+		if(user.shoot_inactive_hand)
+			active_gun.last_fired = world.time
+		else
+			inactive_gun.last_fired = world.time
+	if(user.shoot_inactive_hand && (!inactive_gun?.cell && (!inactive_gun.in_chamber?.ammo && !inactive_gun.current_mag?.current_rounds || inactive_gun.current_mag?.current_rounds && inactive_gun.current_mag.current_rounds <= 0) || inactive_gun?.cell && !inactive_gun.cell?.charge && inactive_gun.cell.charge <= 0)) // Check inactive gun
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:672]: this is a huge mess, do not deal with energy guns here

make a proc in energy gun 
/obj/item/weapon/gun/energy/set_shoot_inactive_hand
And this proc will be adapted for energy gun.

File: code/modules/projectiles/gun_system.dm [hunk 15]
```diff
@@ -652,6 +655,30 @@ User can be passed as null, (a gun reloading itself for instance), so we need to
...
+/obj/item/weapon/gun/proc/set_shoot_inactive_hand(mob/user, firing) // Handles akimbo
+	if(!user)
+		return
+	if(!dual_wield)
+		user.shoot_inactive_hand = FALSE
+		return
+	
+	var/obj/item/weapon/gun/active_gun = user.get_active_held_item()
+	var/obj/item/weapon/gun/inactive_gun = user.get_inactive_held_item()
+	if(firing) // Apply akimbo delay
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:667]: this doesn't need to be in this proc(from l667 to l671). Put that l844

File: code/modules/projectiles/gun_system.dm [hunk 16]
```diff
@@ -652,6 +655,30 @@ User can be passed as null, (a gun reloading itself for instance), so we need to
...
+		if(user.shoot_inactive_hand)
+			active_gun.last_fired = world.time
+		else
+			inactive_gun.last_fired = world.time
+	if(user.shoot_inactive_hand && (!inactive_gun?.cell && (!inactive_gun.in_chamber?.ammo && !inactive_gun.current_mag?.current_rounds || inactive_gun.current_mag?.current_rounds && inactive_gun.current_mag.current_rounds <= 0) || inactive_gun?.cell && !inactive_gun.cell?.charge && inactive_gun.cell.charge <= 0)) // Check inactive gun
+		user.shoot_inactive_hand = FALSE // Shoot from active
+	else if(!user.shoot_inactive_hand && (!active_gun?.cell && (!active_gun.in_chamber?.ammo && !active_gun.current_mag?.current_rounds || active_gun.current_mag?.current_rounds && active_gun.current_mag.current_rounds <= 0) || active_gun?.cell && !active_gun.cell?.charge && active_gun.cell.charge <= 0)) // Check active gun
+		user.shoot_inactive_hand = TRUE // Shoot from inactive
+	else if(firing)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:676]: can you explain to me when does this is usefull?

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_system.dm (hunk 13 above)
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm:658]: Autodoc?

File: code/modules/mob/mob_defines.dm (hunk 10 above)
Comment (MiniMeatwad) [reply code/modules/mob/mob_defines.dm:30]: Autodoc?

File: code/modules/projectiles/gun_system.dm (hunk 16 above)
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm]: It isn't useful, it's shitcode

File: code/modules/projectiles/gun_system.dm (hunk 15 above)
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm]: Done

File: code/modules/projectiles/gun_system.dm (hunk 14 above)
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm]: Removed the energy gun handling in gun system, but I don't know how to set it up for the energy gun now that it's there.

File: code/modules/projectiles/gun_system.dm (hunk 12 above)
Comment (MiniMeatwad) [reply code/modules/projectiles/gun_system.dm]: I don't even know if it's in energy gun code

**Comment by MiniMeatwad** [comment]
There is no interest in this PR, no reply in 8 days. And I was told that you'd rather do it yourself. Either that or I might re-write it in the future.

**Comment by BraveMole** [comment]
> There is no interest in this PR, no reply in 8 days. And I was told that you'd rather do it yourself. Either that or I might re-write it in the future.

You drafted that, and i pinged you 3 TIMES to ask you if you wanted to do finish it?????

**Comment by MiniMeatwad** [comment]
I drafted it because it wasn't suitable for merging and still had bugs, and I missed your last ping. Either that was a discord bug or I skipped it by accident looking through the other miscellaneous pings that people kept doing for inconsequential reasons. I don't recall being pinged in #dev-general about akimbo after september 21st, and I left the comments I had about it on the PR. But the problem with akimbo right now is that I want to know how you want the function adapted for energy gun code since the akimbo checks can't be put in gun_system.

**Comment by MiniMeatwad** [comment]
The earlier commit 61d597d does work without bugs though. But I was just working on akimbo as requested, and I didn't feel up to doing it on my own initiative. Between that and the couple times you dissuaded me from doing it and not replying, I stopped modifying it so as to wait for your reply on what you want me to do.

**Pull request closed without merging**